### ISSUE TITLE
Heap corruption (SIGABRT/SIGSEGV) in eager tf.matmul with rank-4 lhs and broadcast rhs on CPU (oneDNN)

### ISSUE THREAD

**Issue by laolvfan** [tensorflow/tensorflow#123109]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

`tf-nightly 2.22.0-dev20260531`; also reproduces on 2.20.

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 20.04.4 LTS

### Mobile device

N/A

### Python version

Python 3.11.15

### Bazel version

N/A

### GCC/compiler version

N/A

### CUDA/cuDNN version

N/A. Reproduced on CPU with `CUDA_VISIBLE_DEVICES=-1`.

### GPU model and memory

N/A. CPU only.

### Current behavior?

Calling `tf.matmul` in eager mode with a rank-4 left operand of shape `[B, 1, M, K]` and a lower-rank right operand that must be broadcast (e.g. `[K, N]`) corrupts the process heap. The op returns the correct result and shape, but the process then aborts, typically at interpreter teardown, with a glibc heap error such as `corrupted size vs. prev_size`, `free(): invalid pointer`, or `corrupted double-linked list` (exit code 134), and occasionally with SIGSEGV (139).

Because this is memory corruption, the exact symptom varies between runs. With the reproducer below, **8 out of 10 runs abort** on `tf-nightly`; the remaining runs exit cleanly, which is expected for an out-of-bounds write whose effect depends on heap layout.

Setting `TF_ENABLE_ONEDNN_OPTS=0` makes the abort disappear entirely (0 out of 5 runs abort), which points at the oneDNN batch-matmul path.

Conditions required to trigger it (each of these variations is clean):

| lhs | rhs | result |
|---|---|---|
| `[24, 1, 5, 4]` | `[4, 6]` | **corrupts heap** |
| `[24, 1, 4, 4]` | `[4, 4]` | **corrupts heap** |
| `[24, 1, 5, 4]` | `[1, 4, 6]` | **corrupts heap** |
| `[24, 5, 4]` (rank-3 lhs) | `[4, 6]` | OK |
| `[24, 2, 5, 4]` (middle dim ≠ 1) | `[4, 6]` | OK |
| `[24, 1, 5, 4]` | `[24, 1, 4, 6]` (same rank, no broadcast) | OK |
| `[2, 1, 5, 4]` (small batch) | `[4, 6]` | OK |

So the trigger appears to require all of: a rank-4 left operand whose second dimension is 1, a right operand of lower rank that is broadcast against it, and a leading batch dimension of roughly 4 or more. The numerical result is correct in every case; this is an out-of-bounds write, not a wrong-output bug.

### Standalone code to reproduce the issue

```python
import os

os.environ["CUDA_VISIBLE_DEVICES"] = "-1"

import tensorflow as tf

print("TensorFlow:", tf.__version__)

a = tf.zeros([24, 1, 5, 4])
b = tf.zeros([4, 6])
c = tf.matmul(a, b)          # rank-4 lhs, rank-2 rhs broadcast
print("result shape:", c.shape)
print("done")
```

### Relevant log output

```text
TensorFlow: 2.22.0-dev20260531
result shape: (24, 1, 5, 6)
done
corrupted size vs. prev_size
```

Exit code 134 (SIGABRT). Note that both lines of output are printed correctly before the abort.

Running the identical script with `TF_ENABLE_ONEDNN_OPTS=0` prints the same output and exits 0 with no abort.

**Comment by laolvfan** [comment]
I ran into what looks like the same corruption through TFLite differential testing— two independently-generated programs that reach the same rank-4-lhs-with-a-size-1-dim + broadcast-rhs `matmul` shape I already documented, just built up through different preceding ops (`conv2d` in one case, `add` in the other) rather than a literally-shaped `matmul` call.

Confirmed via the same diagnostic I used: `TF_ENABLE_ONEDNN_OPTS=0` eliminates the crash in both cases.

**Case 1** (most reliable repro I found, 8/8 crashes with oneDNN on, 0/5 with it off):

```python
import tensorflow as tf

class Model(tf.keras.Model):
    def call(self, x2):
        w2 = tf.Variable([[3., 4., 1.9], [5.4, 6.5, 2.], [7., 8., 1.]])
        v2 = tf.reshape(x2, (2, 3))
        v3 = tf.nn.relu(w2)
        out = tf.matmul(v2, v3)          # [2, 3]
        out2 = tf.nn.relu(out)
        v3b = tf.reshape(w2, (9, 1, 1, 1))
        added = tf.add(out2, v3b)        # broadcasts to [9, 1, 2, 3]
        y = tf.matmul(added, w2)         # rank-4 lhs w/ size-1 dim x rank-2 rhs -- crashes here
        return y

m = Model()
x2 = tf.constant([1., 2., 3., 4., 5., 6.], shape=[3, 2])
print(m(x2))
```

```text
$ for i in 1 2 3 4 5 6 7 8; do python repro.py >/dev/null 2>&1; echo "run $i -> rc=$?"; done
run 1 -> rc=139   # SIGSEGV
run 2 -> rc=139
run 3 -> rc=139
run 4 -> rc=139
run 5 -> rc=139
run 6 -> rc=139
run 7 -> rc=139
run 8 -> rc=139

$ for i in 1 2 3 4 5; do TF_ENABLE_ONEDNN_OPTS=0 python repro.py >/dev/null 2>&1; echo "run $i -> rc=$?"; done
run 1 -> rc=0
run 2 -> rc=0
run 3 -> rc=0
run 4 -> rc=0
run 5 -> rc=0
```

**Case 2**: a program combining `conv2d` + broadcast + `matmul` (the shape here is `[10,1,1,2]` conv2d output broadcast/multiplied against `[2,2]` constants then fed into `matmul` with a `[2,2]` weight) hits the same corruption, but with a more interesting non-determinism angle: I found near-duplicate variants of the same program (differing only in a trailing `tf.nn.relu` vs `tf.nn.gelu` applied to a *scalar* value at the very end, well after the `matmul` in question) where one variant crashes immediately during eager execution and the other crashes only *after* successfully running through `TFLiteConverter.convert()` — i.e. the corrupted heap metadata doesn't get "discovered" by the allocator until much later, depending on unrelated subsequent allocation activity. Across 8 runs of one such variant I saw crashes at rates similar to what you reported (5/8 here vs. your 8/10), consistent with the same non-deterministic corruption. Same fix: `TF_ENABLE_ONEDNN_OPTS=0` → 0/5 crashes across repeated runs.

I think this is useful mainly as confirmation that the trigger surface is broader than a single literally-shaped `matmul` call — any op sequence that produces a rank-4 tensor with a size-1 batch dim feeding into `matmul` against a lower-rank operand seems to hit it, and the corruption can silently survive well past the original call site (in my case, past an entire `TFLiteConverter.convert()` call) before crashing.

**Comment by Ashutosh0x** [comment]
I believe this is the same root cause as #117700, and #120199 (approved, currently labelled `ready to pull`) should fix it.

In `MklBatchMatMulHelper`, the reshaped dimensions for a broadcast operand are padded with leading `1`s, but `CalculateTFStrides` then computes ordinary non-zero strides for those padded dimensions. oneDNN therefore advances the pointer along a dimension that has no backing data, reading and writing past the end of the buffer. The fix is to force the stride to 0 for any batch dimension of size 1:

```c++
    for (int i = 0; i < ndims_out - 2; ++i) {
      if (lhs_dims[i] == 1) lhs_strides[i] = 0;
      if (rhs_dims[i] == 1) rhs_strides[i] = 0;
    }
```

That matches every condition in your table exactly:

- rank-4 lhs with second dimension `1` — the `lhs_dims[i] == 1` case;
- lower-rank rhs that must be broadcast — creates the padded leading dims;
- `[24, 5, 4]` rank-3 lhs, `[24, 2, 5, 4]` with middle dim ≠ 1, and same-rank `[24, 1, 4, 6]` rhs are all clean, because none of them produce a padded broadcast dimension with a non-zero stride;
- `TF_ENABLE_ONEDNN_OPTS=0` avoiding it entirely, since this code is only on the oneDNN path.

It also explains why the numerical result is always correct: the stride error affects where the kernel reads/writes relative to the buffer, not the arithmetic.

**Caveat on verification:** I ran your reproducer 8 times on `tf-nightly 2.22.0-dev20260726` on Windows and got 0 aborts (result correct each time). I don't think that contradicts the report — Windows' allocator doesn't do glibc's `corrupted size vs. prev_size` style checking, so an out-of-bounds write of this size often lands in slack space undetected. Your Linux 8/10 abort rate is the more meaningful signal. So I'm linking these on code analysis, not on a reproduction.

Might be worth adding `Fixes #123109` to #120199 before it's pulled, if the maintainers agree the diagnosis holds.

**Comment by laolvfan** [comment]
Hi @Ashutosh0x, Thanks for the detailed root-cause analysis and for linking this to #117700/#120199 — the stride explanation matches everything in my trigger-condition table.

I re-tested this on the newest `tf-nightly` build I could get (`2.22.0-dev20260726`), and something changed: neither of my two reproducers crashes anymore.

```text
$ for i in $(seq 1 15); do python matmul_repro.py >/dev/null 2>&1; echo "run $i -> rc=$?"; done
run 1 -> rc=0
...
run 15 -> rc=0   # all 15 clean, same rank-4-lhs/broadcast-rhs matmul as the original report

$ for i in $(seq 1 20); do python case1_repro.py >/dev/null 2>&1; echo "run $i -> rc=$?"; done
run 1 -> rc=0
...
run 20 -> rc=0   # all 20 clean, same reshape+matmul+add+matmul repro from my earlier comment
```

oneDNN is confirmed still enabled in this build. For comparison, the exact same two scripts crashed at high rates on TF 2.20.0 and on `tf-nightly` builds from 2026-05-31 and 2026-07-13.

Neither #120199 nor #117700 shows as merged yet, so it's not clear whether this was actually fixed by something else in the interim or whether it's just no longer being caught. Flagging it in case it's useful for confirming the status of the underlying issue before #120199 lands — and if the diagnosis does hold, it'd be great if the merge could also reference this issue (e.g. `Fixes #123109` alongside #117700), since both are tracking the same root cause.
